identity/src/test/java/com/android/identity/SessionEncryptionInteractionTest.java

```diff
@@ -0,0 +1,107 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.identity;
+
+import androidx.core.util.Pair;
+import androidx.test.filters.SmallTest;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.math.BigInteger;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.util.OptionalLong;
+import java.util.Random;
+
+import co.nstant.in.cbor.CborBuilder;
+import co.nstant.in.cbor.model.Array;
+import co.nstant.in.cbor.model.ByteString;
+import co.nstant.in.cbor.model.DataItem;
+
+public class SessionEncryptionInteractionTest {
+
+    public void testCurve(@Constants.EcCurve int curve) {
+        KeyPair eReaderKeyPair = Util.createEphemeralKeyPair(curve);
+        PublicKey eReaderKeyPublic = eReaderKeyPair.getPublic();
+        PrivateKey eReaderKeyPrivate = eReaderKeyPair.getPrivate();
+        KeyPair eDeviceKeyPair = Util.createEphemeralKeyPair(curve);
+        PublicKey eDeviceKeyPublic = eDeviceKeyPair.getPublic();
+        PrivateKey eDeviceKeyPrivate = eDeviceKeyPair.getPrivate();
+
+        byte[] encodedSessionTranscript = new byte[] {1, 2, 3};
+
+        SessionEncryptionReader sessionEncryptionReader = new SessionEncryptionReader(
+                eReaderKeyPrivate, eReaderKeyPublic, eDeviceKeyPublic, encodedSessionTranscript);
+        sessionEncryptionReader.setSendSessionEstablishment(false);
+        SessionEncryptionDevice sessionEncryptionHolder = new SessionEncryptionDevice(
+                eDeviceKeyPrivate, eReaderKeyPublic, encodedSessionTranscript);
+
+        // have reader generate sessionEstablishment and check holder decryption of sessionEstablishment
+        byte[] mdocRequest = sessionEncryptionReader.encryptMessageToDevice(
+                Util.fromHex(TestVectors.ISO_18013_5_ANNEX_D_DEVICE_REQUEST),
+                OptionalLong.empty());
+        Pair<byte[], OptionalLong> result = sessionEncryptionHolder.decryptMessageFromReader(
+                mdocRequest);
+        Assert.assertFalse(result.second.isPresent());
+        Assert.assertArrayEquals(
+                Util.fromHex(TestVectors.ISO_18013_5_ANNEX_D_DEVICE_REQUEST),
+                result.first);
+
+        // have holder generate deviceResponse and check reader decryption of deviceResponse
+        byte[] deviceResponse = sessionEncryptionHolder.encryptMessageToReader(
+                Util.fromHex(TestVectors.ISO_18013_5_ANNEX_D_DEVICE_RESPONSE),
+                OptionalLong.empty());
+        result = sessionEncryptionReader.decryptMessageFromDevice(deviceResponse);
+        Assert.assertFalse(result.second.isPresent());
+        Assert.assertArrayEquals(
+                Util.fromHex(TestVectors.ISO_18013_5_ANNEX_D_DEVICE_RESPONSE),
+                result.first);
```

Comment:
Suggest to loop over this two times to check that counters are increased correctly. This is also a good place to test that `getNumMessagesEncrypted()` and `getNumMessagesDecrypted()`.